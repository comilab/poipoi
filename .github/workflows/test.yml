name: test

on:
  push:
    branches:
      - 'master'
      - 'develop'
  pull_request:
    types: [opened, synchronize]
    branches:
      - 'master'
      - 'develop'

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: docker-compose
        run: |
          docker-compose -f docker-compose.ci.yml up -d
      - name: setup-php
        run: |
          docker-compose exec -T php composer install --no-interaction
          docker-compose exec -T php yarn install
          docker-compose exec -T php cp .env.example .env
          docker-compose exec -T php php artisan key:generate
          docker-compose exec -T php touch database/database.sqlite
      - name: setup-client
        run: |
          docker-compose exec -T node yarn install
      - name: test-php
        run: docker-compose exec -T php php artisan test
      - name: test-client
        run: docker-compose exec -T node yarn test
