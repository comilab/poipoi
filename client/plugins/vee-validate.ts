import Vue from 'vue'
import { ValidationObserver, ValidationProvider, extend, localize } from 'vee-validate'
import * as rules from 'vee-validate/dist/rules'
import ja from 'vee-validate/dist/locale/ja.json'

for (const [rule, validation] of Object.entries(rules)) {
  extend(rule, {
    ...validation as any
  })
}

localize({
  ja
})
localize('ja')

Vue.component('ValidationObserver', ValidationObserver)
Vue.component('ValidationProvider', ValidationProvider)
