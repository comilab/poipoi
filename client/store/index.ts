import { defineComponent, provide, inject } from 'nuxt-composition-api'
import { InjectionKey } from '@vue/composition-api'

import sessionStore from '~/store/session'
import settingStore from '~/store/setting'
import toastStore from '~/store/toast'

type Store = {
  session: ReturnType<typeof sessionStore>
  setting: ReturnType<typeof settingStore>
  toast: ReturnType<typeof toastStore>
}

const makeStore = () => {
  return {
    session: sessionStore(),
    setting: settingStore(),
    toast: toastStore()
  }
}

export const key: InjectionKey<Store> = Symbol('Store')

export const store: Store = makeStore()

export const StoreProvider = defineComponent({
  setup () {
    provide(key, store)
  },
  render (h) {
    return h('div', this.$slots.default)
  }
})

export const useStore = () => inject(key)!

export const resetStore = () => {
  Object.assign(store, makeStore())
}

export const actions = {
  async nuxtClientInit () {
    try {
      await Promise.all([
        store.session.load(),
        store.setting.load()
      ])
    } catch (error) {
      //
    }
  }
}
