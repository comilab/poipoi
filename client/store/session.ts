import { reactive, toRefs, computed } from 'nuxt-composition-api'
import axios from 'axios'

import User from '~/models/User'

export default function sessionStore () {
  const state = reactive({
    user: null as User|null
  })

  const isLogin = computed(() => !!state.user)

  async function create (query: any) {
    await axios.get('/sanctum/csrf-cookie', {
      baseURL: process.env.apiUrl
    })
    await axios.post('/sessions', {
      ...query,
      device_name: 'poipoi-web'
    })
  }

  async function load () {
    try {
      const { data } = await axios.get('/sessions')
      setUser(new User(data))
    } catch (error) {
      setUser(null)
    }
  }

  async function destroy () {
    try {
      await axios.delete('/sessions/poipoi-web')
    } catch (error) {
      //
    }
    setUser(null)
  }

  function setUser (user: User|null) {
    state.user = user
  }

  return {
    ...toRefs(state),
    isLogin,
    create,
    load,
    destroy,
    setUser
  }
}
