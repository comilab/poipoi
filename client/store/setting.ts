import { reactive, toRefs } from 'nuxt-composition-api'
import axios from 'axios'

import useApi from '~/composables/use-api'
import Setting from '~/models/Setting'

export default function sessionStore () {
  const api = useApi('/settings', Setting)

  const state = reactive({
    data: null as Setting|null
  })

  async function load () {
    const { data } = await axios.get('/settings')
    setData(new Setting(data))
  }

  async function save (data: any) {
    await api.save(data)
    setData(api.item.value!)
  }

  function setData (data: Setting|null) {
    state.data = data
  }

  return {
    ...toRefs(state),
    load,
    save,
    setData
  }
}
