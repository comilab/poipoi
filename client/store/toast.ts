import { reactive, toRefs } from 'nuxt-composition-api'

export type Toast = {
  type: 'success'|'info'|'warning'|'danger'
  message: string
  timeout?: number
}

export default function toastStore () {
  const state = reactive({
    toasts: [] as Toast[]
  })

  function push ({ type, message, timeout = 5000 }: Toast) {
    const toast = {
      type,
      message,
      timeout
    }
    state.toasts.push(toast)
    if (timeout) {
      setTimeout(() => {
        remove(toast)
      }, timeout)
    }
  }

  function remove (toast: Toast) {
    const index = state.toasts.findIndex(t => t === toast)
    if (index >= 0) {
      state.toasts.splice(index, 1)
    }
  }

  return {
    ...toRefs(state),
    push,
    remove
  }
}
