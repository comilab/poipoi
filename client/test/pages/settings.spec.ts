import { fireEvent, render, waitFor, within } from '@testing-library/vue'
import { reactive, Ref } from '@vue/composition-api'
import * as NuxtCompositionApi from 'nuxt-composition-api'
import { validate } from 'vee-validate'
import faker from 'faker'
import VueScrollTo from 'vue-scrollto'

import Settings from '~/pages/settings.vue'
import * as store from '~/store'
import { fakeText, getRules } from '~/test/helper'
import useValidationObserverStub from '~/test/stubs/ValidationObserver'
import settingFactory from '~/test/factories/setting'
import userFactory from '~/test/factories/user'

jest.mock('vue-scrollto')

describe('pages/settings', () => {
  let options: any
  let fetchMock: any

  beforeEach(() => {
    options = {
      stubs: {
        FontAwesomeIcon: true,
        VPopover: true,
        AppTwemojiPicker: true
      }
    }

    store.store.setting.setData(settingFactory.build())
    store.store.session.setUser(userFactory.build())
    jest.spyOn(store.store.setting, 'save').mockResolvedValue()
    jest.spyOn(store.store.session, 'load').mockResolvedValue()
    jest.spyOn(store, 'useStore').mockReturnValue(store.store)

    fetchMock = {
      fetchState: reactive({
        pending: false
      })
    }
    jest.spyOn(NuxtCompositionApi, 'useFetch').mockReturnValue(fetchMock)
  })

  test('renders correctly', () => {
    store.store.setting.setData(settingFactory.build({
      siteTitle: 'site-title',
      siteDescription: 'site-description',
      perPage: 10,
      denyRobot: true,
      enableFeed: true,
      postDefault: {
        denyRobotScope: [],
        enableReaction: true,
        allowedEmojis: [],
        deniedEmojis: [],
        enableTwitterShare: true
      }
    }))
    store.store.session.setUser(userFactory.build({
      name: 'user-name',
      email: 'user@example.com'
    }))
    const { container } = render(Settings, options)

    expect(container).toMatchSnapshot()
  })

  describe('form', () => {
    describe('inputs', () => {
      describe('ユーザ名', () => {
        let rules: any

        beforeEach(() => {
          rules = getRules(Settings, 'ユーザ名', options)!
        })

        test('必須', async () => {
          const result1 = await validate('', rules)
          expect(result1.valid).toBeFalsy()
          expect(result1.failedRules.required).toBeTruthy()

          const result2 = await validate(faker.internet.email(), rules)
          expect(result2.valid).toBeTruthy()
        })
      })

      describe('メールアドレス', () => {
        let rules: any

        beforeEach(() => {
          rules = getRules(Settings, 'メールアドレス', options)!
        })

        test('必須', async () => {
          const result1 = await validate('', rules)
          expect(result1.valid).toBeFalsy()
          expect(result1.failedRules.required).toBeTruthy()

          const result2 = await validate('foo@example.com', rules)
          expect(result2.valid).toBeTruthy()
        })

        test('メールアドレス', async () => {
          const result1 = await validate(faker.random.word(), rules)
          expect(result1.valid).toBeFalsy()
          expect(result1.failedRules.email).toBeTruthy()

          const result2 = await validate('foo@example.com', rules)
          expect(result2.valid).toBeTruthy()
        })
      })

      describe('サイト名', () => {
        let rules: any

        beforeEach(() => {
          rules = getRules(Settings, 'サイト名', options)!
        })

        test('必須', async () => {
          const result1 = await validate('', rules)
          expect(result1.valid).toBeFalsy()
          expect(result1.failedRules.required).toBeTruthy()

          const result2 = await validate(faker.internet.email(), rules)
          expect(result2.valid).toBeTruthy()
        })

        test('32文字まで', async () => {
          const result1 = await validate(fakeText(33), rules)
          expect(result1.valid).toBeFalsy()
          expect(result1.failedRules.max).toBeTruthy()

          const result2 = await validate(fakeText(32), rules)
          expect(result2.valid).toBeTruthy()
        })
      })

      describe('サイト概要', () => {
        let rules: any

        beforeEach(() => {
          rules = getRules(Settings, 'サイト概要', options)!
        })

        test('200文字まで', async () => {
          const result1 = await validate(fakeText(201), rules)
          expect(result1.valid).toBeFalsy()
          expect(result1.failedRules.max).toBeTruthy()

          const result2 = await validate(fakeText(200), rules)
          expect(result2.valid).toBeTruthy()
        })
      })

      describe('1ページあたりの表示件数', () => {
        let rules: any

        beforeEach(() => {
          rules = getRules(Settings, '1ページあたりの表示件数', options)!
        })

        test('必須', async () => {
          const result1 = await validate('', rules)
          expect(result1.valid).toBeFalsy()
          expect(result1.failedRules.required).toBeTruthy()

          const result2 = await validate('1', rules)
          expect(result2.valid).toBeTruthy()
        })

        test('数値', async () => {
          const result1 = await validate('a', rules)
          expect(result1.valid).toBeFalsy()
          expect(result1.failedRules.numeric).toBeTruthy()

          const result2 = await validate('1', rules)
          expect(result2.valid).toBeTruthy()
        })
      })
    })

    describe('submit', () => {
      let invalid: Ref<boolean>

      beforeEach(() => {
        const stub = useValidationObserverStub()
        options.stubs.ValidationObserver = stub.component
        invalid = stub.invalid
      })

      test('入力エラーがあれば送信不可', () => {
        invalid.value = true
        const { getByRole } = render(Settings, options)

        expect(getByRole('button', { name: '保存' })).toHaveAttribute('disabled')
      })

      test('入力エラーがなければ送信可能', () => {
        const { getByRole } = render(Settings, options)

        expect(getByRole('button', { name: '保存' })).not.toHaveAttribute('disabled')
      })

      test('送信中は送信不可', async () => {
        const { getByRole } = render(Settings, options)
        await fireEvent.submit(getByRole('form'))

        expect(getByRole('button', { name: '保存' })).toHaveAttribute('disabled')
      })

      test('送信中はindeterminate = true', async () => {
        const { getByRole } = render(Settings, options)
        expect(within(getByRole('button', { name: '保存' })).queryByRole('status')).not.toBeInTheDocument()

        await fireEvent.submit(getByRole('form'))

        expect(within(getByRole('button', { name: '保存' })).queryByRole('status')).toBeInTheDocument()
      })
    })

    describe('onSubmit', () => {
      test('store.setting.saveを実行', async () => {
        const { getByRole } = render(Settings, options)
        await fireEvent.submit(getByRole('form'))

        expect(store.store.setting.save).toBeCalled()
      })

      test('store.session.loadを実行', async () => {
        const { getByRole } = render(Settings, options)
        await fireEvent.submit(getByRole('form'))

        await waitFor(() => {
          expect(store.store.session.load).toBeCalled()
        })
      })

      test('上部にスクロール', async () => {
        const { getByRole } = render(Settings, options)
        await fireEvent.submit(getByRole('form'))

        expect(VueScrollTo.scrollTo).toBeCalledWith('#top')
      })
    })
  })
})
