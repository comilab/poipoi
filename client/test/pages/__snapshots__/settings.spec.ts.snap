// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/settings renders correctly 1`] = `
<div>
  <div>
    <form
      class="max-w-screen-md mx-auto"
      name="settings"
    >
      <section
        class="rounded bg-white my-8 p-4"
        data-testid="section"
      >
        <h3
          class="text-lg"
          data-testid="title"
        >
          ユーザ設定
        </h3>
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-ユーザ名"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-0"
          >
            ユーザ名
          </label>
          <input
            class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 bg-white p-2"
            data-testid="input"
            id="input-0"
            name="ユーザ名"
            type="text"
          />
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <!---->
          </div>
        </div>
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-メールアドレス"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-1"
          >
            メールアドレス
          </label>
          <input
            class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 bg-white p-2"
            data-testid="input"
            id="input-1"
            name="メールアドレス"
            type="email"
          />
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <!---->
          </div>
        </div>
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-パスワード"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-2"
          >
            パスワード
          </label>
          <input
            class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 bg-white p-2"
            data-testid="input"
            id="input-2"
            name="パスワード"
            type="password"
          />
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <!---->
          </div>
        </div>
      </section>
      <section
        class="rounded bg-white my-8 p-4"
        data-testid="section"
      >
        <h3
          class="text-lg"
          data-testid="title"
        >
          サイト設定
        </h3>
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-サイト名"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-3"
          >
            サイト名
          </label>
          <input
            class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 bg-white p-2"
            data-testid="input"
            id="input-3"
            name="サイト名"
            type="text"
          />
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <div
              class="text-gray-500 text-right mt-1"
              data-testid="counter"
            >
              10 / 32
            </div>
          </div>
        </div>
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-サイト概要"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-4"
          >
            サイト概要
          </label>
          <textarea
            class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 p-2"
            data-testid="input"
            id="input-4"
            name="サイト概要"
            style="height: auto;"
          />
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <div
              class="text-gray-500 text-right mt-1"
              data-testid="counter"
            >
              16 / 200
            </div>
          </div>
        </div>
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-1ページあたりの表示件数"
          min="1"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-5"
          >
            1ページあたりの表示件数
          </label>
          <input
            class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 bg-white p-2"
            data-testid="input"
            id="input-5"
            min="1"
            name="1ページあたりの表示件数"
            type="number"
          />
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <!---->
          </div>
        </div>
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-トップページの検索避け"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-6"
          >
            トップページの検索避け
          </label>
          <div
            class="flex flex-wrap -mx-4 -my-2"
            data-testid="input"
          >
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                name="トップページの検索避け"
                type="radio"
                value="true"
              />
              <span
                class="whitespace-nowrap"
              >
                する
              </span>
            </label>
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                name="トップページの検索避け"
                type="radio"
                value="false"
              />
              <span
                class="whitespace-nowrap"
              >
                しない
              </span>
            </label>
          </div>
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <!---->
          </div>
        </div>
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-RSS配信"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-7"
          >
            RSS配信
          </label>
          <div
            class="flex flex-wrap -mx-4 -my-2"
            data-testid="input"
          >
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                name="RSS配信"
                type="radio"
                value="true"
              />
              <span
                class="whitespace-nowrap"
              >
                配信する
              </span>
            </label>
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                name="RSS配信"
                type="radio"
                value="false"
              />
              <span
                class="whitespace-nowrap"
              >
                配信しない
              </span>
            </label>
          </div>
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <!---->
          </div>
        </div>
      </section>
      <section
        class="rounded bg-white my-8 p-4"
        data-testid="section"
      >
        <h3
          class="text-lg"
          data-testid="title"
        >
          投稿基本設定
        </h3>
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-null"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-8"
          >
            <span>
              デフォルトで検索避けを行う対象
            </span>
            <vpopover-stub
              class="inline-block"
              placement="top"
              popper-options="[object Object]"
              trigger="click hover"
            >
              <button
                class="text-blue-600 cursor-pointer focus_outline-none ml-2"
                type="button"
              >
                <fontawesomeicon-stub
                  icon="[object Object]"
                />
              </button>
            </vpopover-stub>
          </label>
          <div
            class="flex flex-wrap -mx-4 -my-2"
            data-testid="input"
          >
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="checkbox"
                value="public"
              />
              <span
                class="whitespace-nowrap"
              >
                全体公開
              </span>
            </label>
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="checkbox"
                value="password"
              />
              <span
                class="whitespace-nowrap"
              >
                パスワード
              </span>
            </label>
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="checkbox"
                value="nsfw"
              />
              <span
                class="whitespace-nowrap"
              >
                閲覧注意
              </span>
            </label>
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="checkbox"
                value="r18"
              />
              <span
                class="whitespace-nowrap"
              >
                R-18
              </span>
            </label>
          </div>
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <!---->
          </div>
        </div>
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-リアクション"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-9"
          >
            <span>
              リアクション
            </span>
            <vpopover-stub
              class="inline-block"
              placement="top"
              popper-options="[object Object]"
              trigger="click hover"
            >
              <button
                class="text-blue-600 cursor-pointer focus_outline-none ml-2"
                type="button"
              >
                <fontawesomeicon-stub
                  icon="[object Object]"
                />
              </button>
            </vpopover-stub>
          </label>
          <div
            class="flex flex-wrap -mx-4 -my-2"
            data-testid="input"
          >
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                name="リアクション"
                type="radio"
                value="true"
              />
              <span
                class="whitespace-nowrap"
              >
                受け付ける
              </span>
            </label>
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                name="リアクション"
                type="radio"
                value="false"
              />
              <span
                class="whitespace-nowrap"
              >
                受け付けない
              </span>
            </label>
          </div>
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <!---->
          </div>
        </div>
        <div
          class="flex flex-col my-4"
        >
          <label
            class="text-sm font-bold py-2"
          >
            <span>
              リアクションに利用する絵文字
            </span>
            <vpopover-stub
              class="inline-block"
              placement="top"
              popper-options="[object Object]"
              trigger="click hover"
            >
              <button
                class="text-blue-600 cursor-pointer focus_outline-none ml-2"
                type="button"
              >
                <fontawesomeicon-stub
                  icon="[object Object]"
                />
              </button>
            </vpopover-stub>
          </label>
          <div
            class="md_grid gap-4 grid-cols-2"
          >
            <div
              class="flex flex-col"
            >
              <!---->
              <div
                class="mb-2"
              >
                <apptwemojipicker-stub
                  data-testid="twemoji-picker"
                >
                  <button
                    class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none bg-blue-600 text-white hover_bg-blue-700 px-3 py-2 rounded"
                    data-testid="button"
                    type="button"
                  >
                    <div>
                      <fontawesomeicon-stub
                        class="fa-fw mr-2"
                        data-testid="icon"
                        icon="[object Object]"
                      />
                    </div>
                    <div
                      class=""
                      data-testid="text"
                    >
                      利用する絵文字を選択
                    </div>
                  </button>
                </apptwemojipicker-stub>
              </div>
              <div
                class="w-full border rounded px-2"
                data-testid="emojis"
              >
                <div
                  class="text-gray-700 m-2"
                >
                  全ての絵文字を使用
                </div>
              </div>
            </div>
            <div
              class="flex flex-col mt-4 md_mt-0"
            >
              <!---->
              <div
                class="mb-2"
              >
                <apptwemojipicker-stub
                  data-testid="twemoji-picker"
                >
                  <button
                    class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none bg-blue-600 text-white hover_bg-blue-700 px-3 py-2 rounded"
                    data-testid="button"
                    type="button"
                  >
                    <div>
                      <fontawesomeicon-stub
                        class="fa-fw mr-2"
                        data-testid="icon"
                        icon="[object Object]"
                      />
                    </div>
                    <div
                      class=""
                      data-testid="text"
                    >
                      利用しない絵文字を選択
                    </div>
                  </button>
                </apptwemojipicker-stub>
              </div>
              <div
                class="w-full border rounded px-2"
                data-testid="emojis"
              >
                <div
                  class="text-gray-700 m-2"
                >
                  全ての絵文字を使用
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-null"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-10"
          >
            <span>
              Twitterシェア
            </span>
            <vpopover-stub
              class="inline-block"
              placement="top"
              popper-options="[object Object]"
              trigger="click hover"
            >
              <button
                class="text-blue-600 cursor-pointer focus_outline-none ml-2"
                type="button"
              >
                <fontawesomeicon-stub
                  icon="[object Object]"
                />
              </button>
            </vpopover-stub>
          </label>
          <div
            class="flex flex-wrap -mx-4 -my-2"
            data-testid="input"
          >
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="radio"
                value="true"
              />
              <span
                class="whitespace-nowrap"
              >
                表示する
              </span>
            </label>
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="radio"
                value="false"
              />
              <span
                class="whitespace-nowrap"
              >
                表示しない
              </span>
            </label>
          </div>
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <!---->
          </div>
        </div>
      </section>
      <div
        class="text-center"
      >
        <button
          class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none bg-blue-600 text-white hover_bg-blue-700 px-3 py-2 rounded"
          data-testid="button"
          type="submit"
        >
          <div>
            <fontawesomeicon-stub
              class="fa-fw mr-2"
              data-testid="icon"
              icon="[object Object]"
            />
          </div>
          <div
            class=""
            data-testid="text"
          >
            保存
          </div>
        </button>
      </div>
    </form>
  </div>
</div>
`;
